File: src/os/info.c
#include "info.h"

static glr_sysinfo_t __glr_sys_info;

glr_sysinfo_t* glr_sys_info() {
    return &__glr_sys_info;
}

#ifdef GLR_WINDOWS
    void glr_sys_info_init() {
        SYSTEM_INFO info;
        GetSystemInfo(&info);
        
        __glr_sys_info.page_size = info.dwPageSize;
        __glr_sys_info.num_cpus = info.dwNumberOfProcessors;
        __glr_sys_info.huge_page_size = GetLargePageMinimum();    
    }

#else
    #include <stdio.h>
    #include <stdlib.h>
    #include <string.h>
    #include <unistd.h>
    #include <sys/sysinfo.h>

    void glr_sys_info_init() {
        __glr_sys_info.huge_page_size = 0;
        __glr_sys_info.num_cpus = get_nprocs();
        __glr_sys_info.page_size = getpagesize();

        FILE* proc;
        char line[0xff] = { 0 };
        if ((proc = fopen("/proc/meminfo", "r"))) {
            while (fgets(line, sizeof(line) - 1, proc))
                if (strstr(line, "Hugepagesize:"))
                    __glr_sys_info.huge_page_size = strtol(line + 13, 0, 10) * 1024;
            fclose(proc);
        }
    }
#endif

File: glr/src/os/info.rs
use super::*;

pub struct SysInfo {
    pub num_cpus: usize,
    pub page_size: usize,
    pub huge_page_size: usize,
}

lazy_static! {
    pub static ref SYS_INFO: SysInfo = unsafe { get_system_info() };
}

#[cfg(windows)]
unsafe fn get_system_info() -> SysInfo {
    let mut info = core::mem::uninitialized();
    GetSystemInfo(&mut info);

    SysInfo {
        page_size: info.dwPageSize as usize,
        num_cpus: info.dwNumberOfProcessors as usize,
        huge_page_size: GetLargePageMinimum(),
    }
}

#[cfg(unix)]
unsafe fn get_system_info() -> SysInfo {
    let mut huge_page_size = 0;
    let page_size = sysconf(_SC_PAGESIZE) as usize;
    let num_cpus = sysconf(_SC_NPROCESSORS_CONF) as usize;
    
    let file = fopen("/proc/meminfo\0".c_str(), "r\0".c_str());
    if !file.is_null() {
        let mut line = [0; 128];
        let size = (line.len() - 1) as i32;
        while !fgets(line.as_mut_ptr(), size, file).is_null() {
            if !strstr(line.as_ptr(), "Hugepagesize:\0".c_str()).is_null() {
                huge_page_size = (strtol((&line[13..]).as_ptr(), null_mut(), 10) * 1024) as usize;
            }
        }
    }

    SysInfo {
        num_cpus,
        page_size,
        huge_page_size
    }
}

